src/compiler.ts

```diff
@@ -0,0 +1,160 @@
+    let op;
+    let queryPartType = null;
+    let type = this.getType(queryPart);
+    let filter;
+    let filters = [];
+    switch (type) {
+      case 'Object':
+        loop: for (queryPartType in queryPart) {
+          op = queryPart[queryPartType];
+          switch (queryPartType) {
+            case '$eq':
+              filters.push(hi.compare(key, eq, op));
+              break;
+            case '$ne':
+              filters.push(negate(this.compilePart(key, op)));
+              break;
+            case '$in':
+              filters.push(hi.compare(key, includes, op));
+              break;
+            case '$nin':
+              filters.push(negate(hi.compare(key, includes, op)));
+              break;
+            case '$regex':
+              filters.push(this.regexp(this.extractRegexp(op, queryPart['$options']), key));
+              break loop;
+            case '$gte':
+              filters.push(hi.compare(key, gte, op));
+              break;
+            case '$lte':
+              filters.push(hi.compare(key, lte, op));
+              break;
+            case '$gt':
+              filters.push(hi.compare(key, gt, op));
+              break;
+            case '$lt':
+              filters.push(hi.compare(key, lt, op));
+              break;
+            case '$elemMatch':
+              filters.push(flow([get(key), map(hi.and(this._compilePredicates<Query>(op))), some(Boolean)]));
+              break;
+            case '$exists':
+              filters.push(hi.exists(key, op));
+              break;
+            default:
+              throw new Error('No support for: ' + queryPartType);
+          }
+        }
+        break;
+      case 'RegExp':
+        filters.push(this.regexp(queryPart, key));
+        break;
+      case 'Array':
+        switch (key) {
+          case '$or':
+            filters.push(
+              // OR
+              hi.or(this._subQuery(queryPart))
+            );
+            break;
+          case '$and':
+            filters.push(hi.and(this._subQuery(queryPart)));
+            break;
+          default:
+            throw new Error('No support for: ' + key);
+        }
+        break;
+      // primitives
+      case 'Number':
+      case 'Boolean':
+      case 'String':
+      case 'Null':
+        filters.push(hi.check(key, eq(queryPart)));
+        break;
+      default:
+        throw new Error('Unable to parse query + ' + key + ' | ' + queryPart);
+    }
+    if (filters.length > 1) {
+      filter = hi.and(filters);
+    } else {
+      filter = filters[0];
+    }
+    return filter;
+  }
+
+  _subQuery<T extends Query[]>(queries: T): Array<(t: Query) => boolean> {
+    let res = map(this._compilePredicates.bind(this))(queries);
+    // should check if there are bad side effects...
+    for (let i = 0; i < res.length; i++) {
+      if (res[i].length > 1) {
+        res[i] = [hi.and(res[i])];
```

Comment:
Added due to typing errors, passes on test but seems risky to do.
```diff
+ res[i] = [hi.and(res[i])];
- res[i] = hi.and(res[i]);
```

Reply:
We should not change any logic. I assume this logic worked before? If so, the only possible outcome is that something might break now. If there are type errors, it's more likely that the types are incorrect.